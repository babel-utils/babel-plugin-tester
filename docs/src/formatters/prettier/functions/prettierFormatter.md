[**babel-plugin-tester**](../../../../README.md)

***

[babel-plugin-tester](../../../../README.md) / [src/formatters/prettier](../README.md) / prettierFormatter

# Function: prettierFormatter()

> **prettierFormatter**(`code`, `options`?): `string` \| `Promise`\<`string`\>

Defined in: [src/formatters/prettier.ts:24](https://github.com/babel-utils/babel-plugin-tester/blob/03734eaa985470bea60d71fab1aa0d0dbdddae3c/src/formatters/prettier.ts#L24)

A prettier-based formatter used to normalize babel output.

If no `filepath` is given, it will be set to `${cwd}/dummy.js` by
default. This is useful to leverage prettier's extension-based parser
inference (which usually ends up triggering babel).

## Parameters

### code

`string`

The result of a babel transformation that should be formatted.

### options?

`object` & `Partial`\<\{ `config`: `MaybePrettierOptions`; `prettierOptions`: `MaybePrettierOptions`; \}\>

Options expected by the ResultFormatter interface.

## Returns

`string` \| `Promise`\<`string`\>

## See

https://prettier.io/docs/en/options.html#file-path
