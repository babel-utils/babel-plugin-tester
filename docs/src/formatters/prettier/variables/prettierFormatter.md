[**babel-plugin-tester**](../../../../README.md)

***

[babel-plugin-tester](../../../../README.md) / [src/formatters/prettier](../README.md) / prettierFormatter

# Variable: prettierFormatter

> `const` **prettierFormatter**: [`ResultFormatter`](../../../types/type-aliases/ResultFormatter.md)\<\{ `config`: `MaybePrettierOptions`; `prettierOptions`: `MaybePrettierOptions`; \}\>

Defined in: [src/formatters/prettier.ts:24](https://github.com/babel-utils/babel-plugin-tester/blob/4d4ff268cbd4a3f5ae326c51e5487f07121f5c9d/src/formatters/prettier.ts#L24)

A prettier-based formatter used to normalize babel output.

If no `filepath` is given, it will be set to `${cwd}/dummy.js` by
default. This is useful to leverage prettier's extension-based parser
inference (which usually ends up triggering babel).

## See

https://prettier.io/docs/en/options.html#file-path
