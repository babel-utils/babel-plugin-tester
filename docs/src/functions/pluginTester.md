[**babel-plugin-tester**](../../README.md)

***

[babel-plugin-tester](../../README.md) / [src](../README.md) / pluginTester

# Function: pluginTester()

> **pluginTester**(`options?`): `void`

Defined in: [src/index.ts:42](https://github.com/babel-utils/babel-plugin-tester/blob/4d4ff268cbd4a3f5ae326c51e5487f07121f5c9d/src/index.ts#L42)

An abstraction around babel to help you write tests for your babel plugin or
preset.

## Parameters

### options?

[`PluginTesterOptions`](../types/interfaces/PluginTesterOptions.md)

## Returns

`void`
