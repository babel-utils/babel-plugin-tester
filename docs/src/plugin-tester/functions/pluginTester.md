[**babel-plugin-tester**](../../../README.md)

***

[babel-plugin-tester](../../../README.md) / [src/plugin-tester](../README.md) / pluginTester

# Function: pluginTester()

> **pluginTester**(`options`): `void`

Defined in: [src/plugin-tester.ts:129](https://github.com/babel-utils/babel-plugin-tester/blob/03734eaa985470bea60d71fab1aa0d0dbdddae3c/src/plugin-tester.ts#L129)

An abstraction around babel to help you write tests for your babel plugin or
preset.

## Parameters

### options

[`PluginTesterOptions`](../../../types/global/interfaces/PluginTesterOptions.md) = `{}`

## Returns

`void`
