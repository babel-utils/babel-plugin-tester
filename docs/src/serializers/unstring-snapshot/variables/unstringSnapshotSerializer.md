[**babel-plugin-tester**](../../../../README.md)

***

[babel-plugin-tester](../../../../README.md) / [src/serializers/unstring-snapshot](../README.md) / unstringSnapshotSerializer

# Variable: unstringSnapshotSerializer

> `const` **unstringSnapshotSerializer**: [`SnapshotSerializer`](../../../type-aliases/SnapshotSerializer.md)

Defined in: [src/serializers/unstring-snapshot.ts:14](https://github.com/babel-utils/babel-plugin-tester/blob/03734eaa985470bea60d71fab1aa0d0dbdddae3c/src/serializers/unstring-snapshot.ts#L14)

If you're using jest and snapshots, then the snapshot output could have a
bunch of bothersome `\"` to escape quotes because when Jest serializes a
string, it will wrap everything in double quotes.

This snapshot serializer removes these quotes.
