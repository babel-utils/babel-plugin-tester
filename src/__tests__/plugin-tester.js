import fs from 'fs'
import path from 'path'
import assert from 'assert'
import {EOL} from 'os'
import pluginTester from '../plugin-tester'
import prettierFormatter from '../formatters/prettier'
import unstringSnapshotSerializer from '../unstring-snapshot-serializer'
import identifierReversePlugin from './helpers/identifier-reverse-plugin'

expect.addSnapshotSerializer(unstringSnapshotSerializer)

const babel = require('@babel/core')

let errorSpy,
  describeSpy,
  itSpy,
  itOnlySpy,
  itSkipSpy,
  equalSpy,
  transformSpy,
  writeFileSyncSpy,
  pendingTests

const noop = () => {}
// eslint-disable-next-line require-await
const promiseWrap = async fn => fn()
const describeMock = (title, body) => {
  body()
}
const titleTesterMock = (title, testFn) => {
  pendingTests.push(promiseWrap(testFn))
}

const simpleTest = 'var hi = "hey";'

const runPluginTester = async (...args) => {
  pendingTests = []
  try {
    pluginTester(...args)
    return await Promise.all(pendingTests)
  } finally {
    pendingTests = []
  }
}

beforeEach(() => {
  equalSpy = jest.spyOn(assert, 'equal')
  errorSpy = jest.spyOn(console, 'error').mockImplementation(noop)
  describeSpy = jest.spyOn(global, 'describe').mockImplementation(describeMock)
  itSpy = jest.spyOn(global, 'it').mockImplementation(titleTesterMock)
  global.it.only = jest.fn(titleTesterMock)
  global.it.skip = jest.fn(titleTesterMock)
  itOnlySpy = global.it.only
  itSkipSpy = global.it.skip
  transformSpy = jest.spyOn(babel, 'transformAsync')
  writeFileSyncSpy = jest
    .spyOn(fs, 'writeFileSync')
    .mockImplementation(() => {})
})

afterEach(() => {
  equalSpy.mockRestore()
  errorSpy.mockRestore()
  describeSpy.mockRestore()
  itSpy.mockRestore()
  itSkipSpy.mockRestore()
  transformSpy.mockRestore()
  writeFileSyncSpy.mockRestore()
})

test('plugin is required', async () => {
  await expect(runPluginTester()).rejects.toThrowErrorMatchingSnapshot()
})

test('exists early if no tests are supplied', async () => {
  const {plugin} = getOptions()
  await runPluginTester({plugin})
  expect(describeSpy).not.toHaveBeenCalled()
  expect(itSpy).not.toHaveBeenCalled()
})

test('exits early if tests is an empty array', async () => {
  const {plugin} = getOptions()
  await runPluginTester({plugin, tests: []})
  expect(describeSpy).not.toHaveBeenCalled()
  expect(itSpy).not.toHaveBeenCalled()
})

test('accepts a title for the describe block', async () => {
  const title = 'describe block title'
  await runPluginTester(getOptions({title}))
  expect(describeSpy).toHaveBeenCalledWith(title, expect.any(Function))
})

test('calls describe and test for a group of tests', async () => {
  const pluginName = 'supergirl'
  const customTitle = 'some custom title'
  const options = getOptions({
    pluginName,
    tests: [simpleTest, simpleTest, {code: simpleTest, title: customTitle}],
  })
  await runPluginTester(options)
  expect(describeSpy).toHaveBeenCalledTimes(1)
  expect(describeSpy).toHaveBeenCalledWith(
    options.pluginName,
    expect.any(Function),
  )
  expect(itSpy).toHaveBeenCalledTimes(3)
  expect(itSpy.mock.calls).toEqual([
    [`1. ${pluginName}`, expect.any(Function)],
    [`2. ${pluginName}`, expect.any(Function)],
    [`${customTitle}`, expect.any(Function)],
  ])
})

test('tests can be skipped', async () => {
  const {plugin} = getOptions()
  await runPluginTester({plugin, tests: [{skip: true, code: '"hey";'}]})
  expect(itSkipSpy).toHaveBeenCalledTimes(1)
  expect(itSpy).not.toHaveBeenCalled()
})

test('tests can be only-ed', async () => {
  const {plugin} = getOptions()
  await runPluginTester({plugin, tests: [{only: true, code: '"hey";'}]})
  expect(itOnlySpy).toHaveBeenCalledTimes(1)
  expect(itSpy).not.toHaveBeenCalled()
})

test('tests cannot be both only-ed and skipped', async () => {
  const {plugin} = getOptions()
  await expect(
    runPluginTester({
      plugin,
      tests: [{only: true, skip: true, code: '"hey";'}],
    }),
  ).rejects.toThrowErrorMatchingSnapshot()
})

test('default will throw if output changes', async () => {
  const tests = ['var hello = "hi";']
  const options = getOptions({plugin: identifierReversePlugin, tests})
  await snapshotOptionsError(options)
})

test('skips falsy tests', async () => {
  const tests = [simpleTest, undefined, null, simpleTest]
  await runPluginTester(getOptions({tests}))
  expect(itSpy).toHaveBeenCalledTimes(2)
})

test('throws if output is incorrect', async () => {
  const tests = [{code: '"hi";', output: '"hey";'}]
  await snapshotOptionsError(getOptions({tests}))
})

test(`assert throws if there's no code`, async () => {
  const tests = [{}]
  await snapshotOptionsError(getOptions({tests}))
})

test('trims and deindents code and output', async () => {
  const tests = [
    {
      code: `
        var someCode = 'cool';
      `,
      output: `
        var someCode = 'cool';
      `,
    },
  ]
  await runPluginTester(getOptions({tests}))
  expect(equalSpy).toHaveBeenCalledWith(
    `var someCode = 'cool';`,
    `var someCode = 'cool';`,
    expect.any(String),
  )
})

test('accepts an empty output', async () => {
  const tests = [
    {
      code: `var eraseMe = 'junk'`,
      output: '',
    },
  ]
  let errorResponse
  try {
    await runPluginTester(
      getOptions({
        plugin: () => ({
          name: 'cleanup',
          visitor: {
            VariableDeclaration(p) {
              p.remove()
            },
          },
        }),
        tests,
      }),
    )
    errorResponse = false
  } catch (error) {
    errorResponse = true
  }

  expect(errorResponse).toEqual(false)
})

test('can get a code and output fixture that is an absolute path', async () => {
  const tests = [
    {
      fixture: getFixturePath('fixture1.js'),
      outputFixture: getFixturePath('outure1.js'),
    },
  ]
  const error = await runPluginTester(getOptions({tests})).catch(e => e)
  const actual = getFixtureContents('fixture1.js')
  const expected = getFixtureContents('outure1.js')
  expect(error).toMatchObject({
    name: expect.stringMatching(/AssertionError/),
    message: 'Output is incorrect.',
    actual,
    expected,
  })
})

test('can pass with fixture and outputFixture', async () => {
  const tests = [
    {
      fixture: getFixturePath('fixture1.js'),
      outputFixture: getFixturePath('fixture1.js'),
    },
  ]
  await runPluginTester(getOptions({tests}))
})

test('throws error if fixture provided and code changes', async () => {
  const tests = [{fixture: getFixturePath('fixture1.js')}]
  await snapshotOptionsError(
    getOptions({plugin: identifierReversePlugin, tests}),
  )
})

test('can resolve a fixture with the filename option', async () => {
  const tests = [
    {
      fixture: 'fixtures/fixture1.js',
      outputFixture: 'fixtures/outure1.js',
    },
  ]
  const error = await runPluginTester(
    getOptions({filename: __filename, tests}),
  ).catch(e => e)
  const actual = getFixtureContents('fixture1.js')
  const expected = getFixtureContents('outure1.js')
  expect(error).toMatchObject({
    name: expect.stringMatching(/AssertionError/),
    message: 'Output is incorrect.',
    actual,
    expected,
  })
})

test('can pass tests in fixtures relative to the filename', async () => {
  await runPluginTester(
    getOptions({
      filename: __filename,
      fixtures: 'fixtures/fixtures',
      tests: null,
    }),
  )
  expect(describeSpy).toHaveBeenCalledTimes(6)
  expect(itSpy).toHaveBeenCalledTimes(14)

  expect(itSpy.mock.calls).toEqual([
    [`cjs`, expect.any(Function)],
    [`js`, expect.any(Function)],
    [`normal`, expect.any(Function)],
    [`changed`, expect.any(Function)],
    [`fixtureOutputExt`, expect.any(Function)],
    [`jsx support`, expect.any(Function)],
    [`nested a`, expect.any(Function)],
    [`nested b`, expect.any(Function)],
    [`tsx support`, expect.any(Function)],
    [`typescript`, expect.any(Function)],
    [`unchanged`, expect.any(Function)],
    [`nested with option`, expect.any(Function)],
    [`nested without option`, expect.any(Function)],
    [`without output file`, expect.any(Function)],
  ])
})

test('can fail tests in fixtures at an absolute path', async () => {
  const error = await runPluginTester(
    getOptions({
      plugin: identifierReversePlugin,
      tests: null,
      fixtures: getFixturePath('failing-fixtures'),
    }),
  ).catch(e => e)
  expect(error.message).toMatchSnapshot()
})

test('creates output file for new tests', async () => {
  await runPluginTester(
    getOptions({
      filename: __filename,
      fixtures: 'fixtures/creates-output-file',
      tests: null,
    }),
  )

  expect(writeFileSyncSpy).toHaveBeenCalledWith(
    path.join(__dirname, './fixtures/creates-output-file/fixture/output.js'),
    `console.log('hello');`,
  )
  expect(writeFileSyncSpy).toHaveBeenCalledTimes(1)
})

test('uses the fixture filename in babelOptions', async () => {
  const fixture = getFixturePath('fixture1.js')
  const tests = [
    {
      fixture,
      outputFixture: getFixturePath('fixture1.js'),
    },
  ]
  await runPluginTester(getOptions({tests}))
  expect(transformSpy).toHaveBeenCalledTimes(1)
  expect(transformSpy).toHaveBeenCalledWith(
    expect.any(String),
    expect.objectContaining({
      filename: fixture,
    }),
  )
})

test('allows for a test babelOptions can provide a filename', async () => {
  const filename = getFixturePath('outure1.js')
  const tests = [
    {
      babelOptions: {filename},
      fixture: getFixturePath('fixture1.js'),
      outputFixture: getFixturePath('fixture1.js'),
    },
  ]
  await runPluginTester(getOptions({tests}))
  expect(transformSpy).toHaveBeenCalledTimes(1)
  expect(transformSpy).toHaveBeenCalledWith(
    expect.any(String),
    expect.objectContaining({
      filename,
    }),
  )
})

test('can provide a test filename for code strings', async () => {
  const filename = getFixturePath('outure1.js')
  const tests = [{babelOptions: {filename}, code: simpleTest}]
  await runPluginTester(getOptions({tests}))
  expect(transformSpy).toHaveBeenCalledTimes(1)
  expect(transformSpy).toHaveBeenCalledWith(
    expect.any(String),
    expect.objectContaining({
      filename,
    }),
  )
})

test('works with versions of babel without `.transformSync` method', async () => {
  const tests = [simpleTest]
  const oldBabel = {
    transform: babel.transform,
  }
  const transformSyncSpy = jest.spyOn(oldBabel, 'transform')
  await runPluginTester(
    getOptions({
      babel: oldBabel,
      filename: __filename,
      fixtures: 'fixtures/fixtures',
      tests,
    }),
  )
  expect(transformSpy).not.toHaveBeenCalled()
  expect(transformSyncSpy).toHaveBeenCalledTimes(15)
})

test('can provide plugin options', async () => {
  const tests = [simpleTest]
  const pluginOptions = {
    optionA: true,
  }
  await runPluginTester(getOptions({tests, pluginOptions}))
  expect(transformSpy).toHaveBeenCalledTimes(1)
  expect(transformSpy).toHaveBeenCalledWith(
    expect.any(String),
    expect.objectContaining({
      plugins: expect.arrayContaining([
        [
          expect.any(Function),
          expect.objectContaining({
            optionA: true,
          }),
        ],
      ]),
    }),
  )
})

test('can overwrite plugin options at test level', async () => {
  const pluginOptions = {
    optionA: false,
  }
  const tests = [{code: simpleTest, pluginOptions}]
  await runPluginTester(
    getOptions({
      tests,
      pluginOptions: {
        optionA: true,
      },
    }),
  )
  expect(transformSpy).toHaveBeenCalledTimes(1)
  expect(transformSpy).toHaveBeenCalledWith(
    expect.any(String),
    expect.objectContaining({
      plugins: expect.arrayContaining([
        [
          expect.any(Function),
          expect.objectContaining({
            optionA: false,
          }),
        ],
      ]),
    }),
  )
})

test('assert throws if snapshot and output are both provided', async () => {
  const tests = [{code: simpleTest, output: 'anything', snapshot: true}]
  await snapshotOptionsError(getOptions({tests}))
})

test('snapshot option can be derived from the root config', async () => {
  const tests = [{code: simpleTest, output: 'anything'}]
  await snapshotOptionsError(getOptions({snapshot: true, tests}))
})

test('assert throws if code is unchanged + snapshot enabled', async () => {
  const tests = [simpleTest]
  await snapshotOptionsError(getOptions({snapshot: true, tests}))
})

test('takes a snapshot', async () => {
  // this one is kinda tricky... At first I thought I'd mock toMatchSnapshot
  // but then I realized that we can actually use it to our advantage in
  // this case. We actually _do_ take a snapshot and that makes our test
  // work pretty well soooooo... 😀
  const tests = [simpleTest]
  await runPluginTester(
    getOptions({snapshot: true, tests, plugin: identifierReversePlugin}),
  )
})

test('can provide an object for tests', async () => {
  const firstTitle = 'first title'
  const secondTitle = 'second title'
  const tests = {
    [firstTitle]: simpleTest,
    [secondTitle]: {
      code: simpleTest,
    },
  }
  await runPluginTester(getOptions({tests}))
  expect(equalSpy).toHaveBeenCalledTimes(2)
  expect(equalSpy.mock.calls).toEqual([
    [simpleTest, simpleTest, expect.any(String)],
    [simpleTest, simpleTest, expect.any(String)],
  ])
  expect(itSpy.mock.calls).toEqual([
    [firstTitle, expect.any(Function)],
    [secondTitle, expect.any(Function)],
  ])
})

test('can capture errors with true', () => testPluginError(true))

test('can capture errors with Error constructor', () =>
  testPluginError(SyntaxError))

test('can capture errors with string', () => testPluginError('test message'))

test('can capture errors with regex', () => testPluginError(/mess/))

test('can capture errors with function', () =>
  testPluginError(
    err => /mess/.test(err.message) && err instanceof SyntaxError,
  ))

test(`throws error when function doesn't return true`, () =>
  snapshotPluginError(() => false))

test('throws error when error expected but no error thrown', () =>
  snapshotPluginError(true, {
    plugin: () => ({}),
  }))

test('throws error if there is a problem parsing', async () => {
  let error
  try {
    await runPluginTester(
      getOptions({
        tests: [`][fkfhgo]fo{r`],
        babelOptions: {filename: __filename},
      }),
    )
  } catch (e) {
    error = e
  }
  expect(error.constructor).toBe(SyntaxError)
  expect(error.message).toContain('Unexpected token (1:0)')
})

test(`throws an error if babelrc is true with no filename`, () => {
  const tests = ['"use strict";']
  snapshotOptionsError(getOptions({tests, babelOptions: {babelrc: true}}))
})

test('runs test setup function', async () => {
  const setupSpy = jest.fn()
  const tests = [{code: simpleTest, setup: setupSpy}]
  await runPluginTester(getOptions({tests}))
  expect(setupSpy).toHaveBeenCalledTimes(1)
})

test('runs test teardown function', async () => {
  const teardownSpy = jest.fn()
  const tests = [{code: simpleTest, teardown: teardownSpy}]
  await runPluginTester(getOptions({tests}))
  expect(teardownSpy).toHaveBeenCalledTimes(1)
})

test('setup can return a teardown function', async () => {
  const teardownSpy = jest.fn()
  const setupSpy = jest.fn(() => teardownSpy)
  const tests = [{code: simpleTest, setup: setupSpy}]
  await runPluginTester(getOptions({tests}))
  expect(teardownSpy).toHaveBeenCalledTimes(1)
})

test('function resolved from setup promise used for teardown', async () => {
  const teardownSpy = jest.fn()
  const setupSpy = jest.fn(() => Promise.resolve(teardownSpy))
  const tests = [{code: simpleTest, setup: setupSpy}]
  await runPluginTester(getOptions({tests}))
  expect(teardownSpy).toHaveBeenCalledTimes(1)
})

test('error logged and thrown if setup throws', async () => {
  const errorToThrow = new Error('blah')
  const setupSpy = jest.fn(() => {
    throw errorToThrow
  })
  const tests = [{code: simpleTest, setup: setupSpy}]
  const errorThrown = await runPluginTester(getOptions({tests})).catch(e => e)
  expect(errorThrown).toBe(errorToThrow)
  expect(errorSpy).toHaveBeenCalledWith(
    expect.stringMatching(/problem.*setup/i),
  )
})

test('error logged and thrown if teardown throws', async () => {
  const errorToThrow = new Error('blah')
  const teardownSpy = jest.fn(() => {
    throw errorToThrow
  })
  const tests = [{code: simpleTest, teardown: teardownSpy}]
  const errorThrown = await runPluginTester(getOptions({tests})).catch(e => e)
  expect(errorThrown).toBe(errorToThrow)
  expect(errorSpy).toHaveBeenCalledWith(
    expect.stringMatching(/problem.*teardown/i),
  )
})

test('allows formatting the result', async () => {
  const formatResultSpy = jest.fn(r => r)
  await runPluginTester(
    getOptions({
      filename: __filename,
      tests: [{code: simpleTest, formatResult: formatResultSpy}],
    }),
  )
  expect(formatResultSpy).toHaveBeenCalledTimes(1)
  expect(formatResultSpy).toHaveBeenCalledWith(simpleTest, {
    filename: __filename,
  })
})

test('allows formatting fixtures results', async () => {
  const formatResultSpy = jest.fn(r => r)
  await runPluginTester(
    getOptions({
      fixtures: getFixturePath('fixtures'),
      formatResult: formatResultSpy,
    }),
  )
  expect(formatResultSpy).toHaveBeenCalledTimes(15)
})

test('works with a formatter adding a empty line', async () => {
  // Simulate prettier adding an empty line at the end
  const formatResultSpy = jest.fn(r => `${r.trim()}\n\n`)
  await runPluginTester(
    getOptions({
      fixtures: getFixturePath('fixtures'),
      formatResult: formatResultSpy,
    }),
  )
  expect(formatResultSpy).toHaveBeenCalledTimes(15)
})

test('prettier formatter supported', async () => {
  await runPluginTester(
    getOptions({
      formatResult: prettierFormatter,
      tests: {
        formatted: {
          code: `
            console.log(  "hey"  )
          `,
          output: `
            console.log('hey')
          `,
        },
      },
    }),
  )
})

test('gets options from options.json files when using fixtures', async () => {
  const optionRootFoo = jest.fn()
  const optionFoo = jest.fn()
  const optionBar = jest.fn()
  const pluginWithOptions = jest.fn(() => {
    return {
      visitor: {
        Program(programPath, state) {
          if (state.opts.rootFoo === 'rootBar') {
            optionRootFoo()
          }
          if (state.opts.foo === 'bar') {
            optionFoo()
          }
          if (state.opts.bar === 'baz') {
            optionBar()
          }
        },
      },
    }
  })

  await runPluginTester(
    getOptions({
      plugin: pluginWithOptions,
      fixtures: getFixturePath('fixtures'),
    }),
  )

  expect(optionRootFoo).toHaveBeenCalledTimes(14)
  expect(optionFoo).toHaveBeenCalledTimes(2)
  expect(optionBar).toHaveBeenCalledTimes(1)
})

test('appends to root plugins array', async () => {
  const optionRootFoo = jest.fn()
  const optionFoo = jest.fn()
  const optionBar = jest.fn()
  const pluginWithOptions = jest.fn(() => {
    return {
      visitor: {
        Program(programPath, state) {
          if (state.opts.rootFoo === 'rootBar') {
            optionRootFoo()
          }
          if (state.opts.foo === 'bar') {
            optionFoo()
          }
          if (state.opts.bar === 'baz') {
            optionBar()
          }
        },
      },
    }
  })
  const programVisitor = jest.fn()
  const otherPlugin = () => {
    return {
      visitor: {
        Program: programVisitor,
      },
    }
  }

  await runPluginTester(
    getOptions({
      plugin: pluginWithOptions,
      fixtures: getFixturePath('fixtures'),
      babelOptions: {
        plugins: [otherPlugin],
      },
    }),
  )

  expect(optionRootFoo).toHaveBeenCalledTimes(14)
  expect(optionFoo).toHaveBeenCalledTimes(2)
  expect(optionBar).toHaveBeenCalledTimes(1)
  expect(programVisitor).toHaveBeenCalledTimes(15)
})

test('endOfLine - default', async () => {
  await runPluginTester(
    getOptions({
      tests: [
        {
          code: 'var foo = "";\nvar bar = "";',
          output: 'var foo = "";\nvar bar = "";',
        },
      ],
    }),
  )
})

test('endOfLine - Unix', async () => {
  await runPluginTester(
    getOptions({
      endOfLine: 'lf',
      tests: [
        {
          code: 'var foo = "";\nvar bar = "";',
          output: 'var foo = "";\nvar bar = "";',
        },
      ],
    }),
  )
})

test('endOfLine - Windows', async () => {
  await runPluginTester(
    getOptions({
      endOfLine: 'crlf',
      tests: [
        {
          code: 'var foo = "";\nvar bar = "";',
          output: 'var foo = "";\r\nvar bar = "";',
        },
      ],
    }),
  )
})

test('endOfLine - auto', async () => {
  await runPluginTester(
    getOptions({
      endOfLine: 'auto',
      tests: [
        {
          code: 'var foo = "";\nvar bar = "";',
          output: `var foo = "";${EOL}var bar = "";`,
        },
      ],
    }),
  )
})

test('endOfLine - preserve', async () => {
  await runPluginTester(
    getOptions({
      endOfLine: 'preserve',
      tests: [
        {
          code: 'var foo = "";\r\nvar bar = "";',
          output: `var foo = "";\r\nvar bar = "";`,
        },
      ],
    }),
  )
})

test('endOfLine - preserve - no linefeed', async () => {
  await runPluginTester(
    getOptions({
      endOfLine: 'preserve',
      tests: [
        {
          code: 'var foo = "";',
          output: 'var foo = "";',
        },
      ],
    }),
  )
})

test('endOfLine - invalid option', async () => {
  const errorThrown = await runPluginTester(
    getOptions({
      endOfLine: 'invalid',
      tests: [
        {
          code: 'var foo = "";',
          output: 'var foo = "";',
        },
      ],
    }),
  ).catch(e => e)

  expect(errorThrown.message).toBe("Invalid 'endOfLine' value")
})

function getOptions(overrides) {
  return {
    pluginName: 'captains-log',
    plugin: () => ({name: 'captains-log', visitor: {}}),
    tests: [simpleTest],
    ...overrides,
  }
}

function getFixtureContents(fixture) {
  const fullPath = getFixturePath(fixture)
  return fs.readFileSync(fullPath, 'utf8').trim()
}

function getFixturePath(fixture = '') {
  return path.join(__dirname, 'fixtures', fixture)
}

async function snapshotOptionsError(options) {
  let error
  try {
    await runPluginTester(options)
  } catch (e) {
    error = e
  }
  expect(error.message).toMatchSnapshot()
}

async function snapshotPluginError(error, overrides) {
  let errorThrown
  try {
    await testPluginError(error, overrides)
  } catch (e) {
    errorThrown = e
  }
  expect(errorThrown.message).toMatchSnapshot()
}

function testPluginError(error, overrides) {
  return runPluginTester(
    getOptions({
      plugin: () => {
        throw new SyntaxError('test message')
      },
      tests: [
        {
          code: simpleTest,
          error,
        },
      ],
      ...overrides,
    }),
  )
}
