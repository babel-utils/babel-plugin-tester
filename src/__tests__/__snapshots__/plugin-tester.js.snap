// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assert throws if code is unchanged + snapshot enabled 1`] = `Code was unmodified but attempted to take a snapshot. If the code should not be modified, set \`snapshot: false\``;

exports[`assert throws if snapshot and output are both provided 1`] = `\`output\` cannot be provided with \`snapshot: true\``;

exports[`assert throws if there's no code 1`] = `A string or object with a \`code\` or \`fixture\` property must be provided`;

exports[`can fail tests in fixtures at an absolute path 1`] = `actual output does not match output.js`;

exports[`default will throw if output changes 1`] = `Expected output to not change, but it did`;

exports[`plugin is required 1`] = `plugin is a required parameter.`;

exports[`snapshot option can be derived from the root config 1`] = `\`output\` cannot be provided with \`snapshot: true\``;

exports[`takes a snapshot: 1. captains-log 1`] = `

var hi = "hey";

      ↓ ↓ ↓ ↓ ↓ ↓

var ih = "hey";

`;

exports[`tests cannot be both only-ed and skipped 1`] = `Cannot enable both skip and only on a test`;

exports[`throws an error if babelrc is true with no filename 1`] = `babelrc set to true, but no filename specified in babelOptions`;

exports[`throws error if fixture provided and code changes 1`] = `Expected output to not change, but it did`;

exports[`throws error when error expected but no error thrown 1`] = `Expected to throw error, but it did not.`;

exports[`throws error when function doesn't return true 1`] = `[BABEL] unknown: test message (While processing: "base$0")`;

exports[`throws if output is incorrect 1`] = `Output is incorrect.`;
